ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/retriever/SimpleSidecarRetriever.java

```diff
@@ -145,63 +154,57 @@ private Optional<ConnectedPeer> findBestMatchingPeer(
+  private Stream<ConnectedPeer> findMatchingPeers(
       final RetrieveRequest request, final RequestTracker ongoingRequestsTracker) {
     return connectedPeers.values().stream()
         .filter(peer -> peer.isCustodyFor(request.columnId))
-        .filter(peer -> ongoingRequestsTracker.hasAvailableRequests(peer.nodeId))
-        .toList();
+        .filter(peer -> ongoingRequestsTracker.hasAvailableRequests(peer.nodeId));
   }
 
   private void disposeCompletedRequests() {
-    final Iterator<Map.Entry<DataColumnSlotAndIdentifier, RetrieveRequest>> pendingIterator =
-        pendingRequests.entrySet().iterator();
-    while (pendingIterator.hasNext()) {
-      final Map.Entry<DataColumnSlotAndIdentifier, RetrieveRequest> pendingEntry =
-          pendingIterator.next();
-      final RetrieveRequest pendingRequest = pendingEntry.getValue();
-      if (pendingRequest.result.isDone()) {
-        pendingIterator.remove();
-        if (pendingRequest.activeRpcRequest != null) {
-          pendingRequest.activeRpcRequest.promise().cancel(true);
-        }
-      }
-    }
+    pendingRequests
+        .entrySet()
+        .removeIf(
+            pendingEntry -> {
+              final RetrieveRequest pendingRequest = pendingEntry.getValue();
+              if (pendingRequest.result.isDone()) {
+                if (pendingRequest.activeRpcRequest != null) {
+                  pendingRequest.activeRpcRequest.promise().cancel(true);
+                }
+                return true;
+              }
+              return false;
+            });
   }
 
   private void nextRound() {
-    final List<RequestMatch> matches = matchRequestsAndPeers();
-    for (final RequestMatch match : matches) {
-      if (match.request.activeRpcRequestSet.compareAndSet(false, true)) {
-        final SafeFuture<DataColumnSidecar> reqRespPromise =
-            reqResp.requestDataColumnSidecar(match.peer.nodeId, match.request.columnId);
-        match.request().onPeerRequest(match.peer().nodeId);
-        match.request.activeRpcRequest =
-            new ActiveRequest(
-                reqRespPromise.whenComplete(
-                    (sidecar, err) -> reqRespCompleted(match.request, sidecar)),
-                match.peer);
-      }
-    }
+    disposeCompletedRequests();
 
-    final long activeRequestCount =
-        pendingRequests.values().stream().filter(r -> r.activeRpcRequest != null).count();
-    LOG.trace(
-        "SimpleSidecarRetriever.nextRound: completed: {}, errored: {},  total pending: {}, active pending: {}, new active: {}, number of custody peers: {}",
-        retrieveCounter,
-        errorCounter,
-        pendingRequests.size(),
-        activeRequestCount,
-        matches.size(),
-        gatherAvailableCustodiesInfo());
+    final long activatedMatches =
+        matchRequestsAndPeers()
+            .map(this::activateMatchedRequest)
+            .filter(activated -> activated)
+            .count();
```

Comment:
also only in the trace...

Reply:
we could have something like
```java
var stream = matchRequestsAndPeers()
            .map(this::activateMatchedRequest);
            
if(traceIsActive) {
            long activatedMatches =
        stream
            .filter(activated -> activated)
            .count();
} else {
    stream.forEach( __ -> {});
}
```

But seems to me an overcomplication